daemon/status.go

```diff
@@ -32,9 +34,305 @@ import (
 )
 
 const (
+	// collectStatusInterval is how often statuses will be collected
 	collectStatusInterval = 5 * time.Second
+
+	// staleTimeout is how stale a status can be before we report an error
+	staleTimeout = 1 * collectStatusInterval
 )
 
+type getHealthz struct {
+	daemon *Daemon
+}
+
+func NewGetHealthzHandler(d *Daemon) GetHealthzHandler {
+	return &getHealthz{daemon: d}
+}
+
+func (h *getHealthz) Handle(params GetHealthzParams) middleware.Responder {
+	d := h.daemon
+	sr := d.daemonStatus.getStatus()
+	return NewGetHealthzOK().WithPayload(&sr)
+}
+
+// daemonStatus is a simple holder of collected statuses. The lock protects
+// direct read/write access to the struct, subcomponents are not guarded by it.
+type daemonStatus struct {
+	lock.RWMutex
+
+	controllers controller.Manager
+
+	controllerStatus   *models.ControllerStatuses
+	controllerStatusTS time.Time
+
+	lockStatusTS time.Time
+
+	kvStoreStatus   *models.Status
+	kvStoreStatusTS time.Time
+
+	containerRuntimeStatus   *models.Status
+	containerRuntimeStatusTS time.Time
+
+	k8sStatus   *models.K8sStatus
+	k8sStatusTS time.Time
+
+	ipamStatus   *models.IPAMStatus
+	ipamStatusTS time.Time
+
+	nodeMonitorStatus   *models.MonitorStatus
+	nodeMonitorStatusTS time.Time
+
+	clusterStatus   *models.ClusterStatus
+	clusterStatusTS time.Time
+
+	healthStatus   *models.Status
+	healthStatusTS time.Time
+
+	l7ProxyStatus   *models.ProxyStatus
+	l7ProxyStatusTS time.Time
+}
+
+// startStatusCollector spawns all the controllers used to collect subsystem
+// statuses for a daemon object d
+func (dss *daemonStatus) startStatusCollector(d *Daemon) {
```

Comment:
You are passing the `daemon` only because of the `IPAM`? Don't he have every single component in packages?  :disappointed: